### PR TITLE
Maintenance Speedup + Small Refactoring

### PR DESCRIPTION
**Pull request by maierlars**
### Scope & Purpose
First compute the shard map outside of the loop. Otherwise we end up with a quadratic algorithm.

Furthermore, while I'm at it, I replaced some adhoc velocypack with some proper c++ data structure.

### CODE REVIEW

File: arangod/Cluster/Maintenance.cpp [hunk 1]
```diff
@@ -1959,7 +1969,7 @@ arangodb::Result arangodb::maintenance::reportInCurrent(
       return replication::Version::ONE;
     });
 
-    VPackBuilder shardMap;
+    ShardMapView shardMap;
```
Comment (mchacki) [inline arangod/Cluster/Maintenance.cpp:1972]:
```suggestion
    // NOTE: shardMap only stays valid as long as plan is not deleted. Do not return it.
    ShardMapView shardMap;
```

File: arangod/Cluster/Maintenance.cpp [hunk 2]
```diff
@@ -1054,19 +1063,19 @@ arangodb::Result arangodb::maintenance::diffPlanLocal(
           continue;
         }
 
+        auto const shardMap = getShardMap(plan);  // plan shards -> servers
```
Comment (mchacki) [inline arangod/Cluster/Maintenance.cpp:1066]:
```suggestion
        // NOTE: shardMap only stays valid as long as plan is not deleted. Do not return it.
        auto const shardMap = getShardMap(plan);  // plan shards -> servers
```

**Review by mchacki — CHANGES_REQUESTED** [review]
Technical: Very good :+1:

Please add a CHANGELOG entry.
For consistency with 3.11 you can take:


```
* Improved the time required to create a new Collection in a database with
  hundreds of collections. This also improves times for indexes and dropping
  of collections.
```

**Review by mchacki — APPROVED** [review]
LGTM!

**Review by jsteemann — APPROVED** [review]
LGTM

**Pull request merged**